### ISSUE TITLE
Experiment with student model parameters

### ISSUE THREAD

**Issue by gregtatum** [mozilla/translations#894]
In [Ludicrously Fast Neural Machine Translation](https://aclanthology.org/D19-5632.pdf), they test a variety of decoder configurations for faster models.

<img width="744" alt="Screenshot of Table 3: Configuration of student models and submissions" src="https://github.com/user-attachments/assets/d9591807-7f9d-42aa-83f8-dd3f6dc58b9c">

In #174 @eu9ene showed that a larger model helps improve the COMET score for `en-ru` by +2.9, which is pretty significant.

(Edit: I changed from en-ru to en-lt)

I'd like to test the parameters a bit more, as these changes are impactful in terms of quality, but also affect the performance of the model. The paper tested parameters on `en-de`, but our training of `en-ru` has struggled to gain the same amount of COMET with the same architecture. Rather than testing `en-ru`, I'll do a clean run on `en-lt` as it had a pretty low COMET score, and also features a much more varied morphology for the language due to its declension system. The idea is that the results will scale to other Balto-Slavic languages.

I'm shortening the labels in the table a bit:
```
dec-depth: depth
dim-emb: emb
transformer-dim-ffn: ffn
```

| COMET        | vs tiny | wps    | depth   | emb   | ffn    | Name            |
| -----------  | ------- | ------ | ------- | ----- | ------ | --------------- |
| `86.67`      |  -      | 1049   | `2`     | `256` | `1536` | decoder-tiny    |
| `88.78`      | `+2.11` |  406   | `2`     | `512` | `2048` | decoder-base    |
| `86.94`      | `+0.27` |  920   | `3`     | `256` | `1536` | decoder-depth-3 |
| `87.84`      | `+1.17` |  603   | `6`     | `256` | `1536` | decoder-depth-6 |
| `86.88`      | `+0.21` |  916   | `2`     | `256` | `2048` | decoder-ffn-bigger |
| `88.47`      | `+1.80` |  463   | `2`     | `512` | `1536` | decoder-emb-bigger |

[Here is a more details spreadsheet of the speed results.](https://docs.google.com/spreadsheets/d/1U2C4RqJXBqIMKvl6tqcyn8dvtn7bfvb7AuPHVW7JZ1g/edit?gid=0#gid=0)

## Links
 * [spring-2024 en-lt task group](https://firefox-ci-tc.services.mozilla.com/tasks/Eb29irNMT-efiBf3vh2__Q)

**Comment by eu9ene** [comment]
I suggest using a different language pair for this experiment. en-ru was trained from a super convoluted branch "release_no_priors" where I had to change the graph by adding an extra step for alignments to do some bug fixes and not retrain everything from scratch. It's far behind main and doesn't have the latest W&B fixes, so I don't want to run any more experiments from the "release" based branches. If we switch to main, the graph will not be compatible, so we'll have to at least rerun the alignments step and reuse some other tasks using "existing_tasks". With all that it's a lot easier to run some other language pair we struggle with from main where we can reuse the tasks from release, for example, en-lt.

On another note, this looks like a hyperparameter search that we can do manually, but there are tools to automate that we might explore in the future.

**Comment by gregtatum** [comment]
Ok, en-lt sounds like a great choice. I read a bit more on it and it's got a lot of qualitative feedback in #756.

**Comment by gregtatum** [comment]
Lithuanian has a similar use of declensions: https://en.wikipedia.org/wiki/Lithuanian_declension

**Comment by gregtatum** [comment]
I've got the first one started, and will wait until it gets to the student step before kicking off the rest:

https://firefox-ci-tc.services.mozilla.com/tasks/groups/Wxvkl1ruQkCj6URG6oIuuQ

The configs are each done on a commit-by-commit level:
https://github.com/mozilla/translations/commits/dev-en-lt-decoder-size/

**Comment by gregtatum** [comment]
They are all in student training now on [the dashboard](https://gregtatum.github.io/taskcluster-tools/src/training/?taskGroupIds=O3KtCsLKTHSRyvarD0OmkA%2CMb6nq4IiRx2ZQALGOl506w%2CF4lNZ2k6T7mr6wyNhdNpqg%2CCP2Xt9qhScW0ByzsFMTGng%2CGoDGckYuSuS9-eJbwx9a9g%2CUBduTJojQDWmoedTxg74ww%2CAuR23z6EQMeZrNM398e10A%2CV4rRhXt6Te-g_004KEiEUQ%2CFiSNJUysQKqf0HDZZ7uBAw%2CQb2MenUpRE-dyR8owan07Q%2CPy4okXqaTzKG4sFYqi9Nqg%2CI5x8d1_RS5avEFpYbEPdmw%2CL4LaCVDQT_6nwt7tMaCSOQ%2CTaeCdUs5Rqq7w1Tbf1PShQ&showAll=false&hidden=CGdli4TrRCu_FyLy3bsmcg%2CK_6r19C1Q-aAjFVS_VP21Q%2Ccfo9K-R8QtmuE27UKQ4BbQ&dashboardName=Experiment+Dashboard). They are all the ones named `decoder-*`.

**Comment by gregtatum** [comment]
So I misread the paper a bit when it was talking about decoders, and the ffn and embedding size affect both decoder and encoder equally. The decoder depth is the only parameter changed that affects the decoder. I'm updating my experiment notes accordingly.

**Comment by ZJaume** [comment]
I think `transformer-dim-ffn` applies for both. But since the decoder is `ssru`, which is a recurrent network, the params that are applied to the decoder should be the `s2s` ones. So I guess to change the feed-forward size of the decoder, `dim-rnn` has to be used?
